gallery/bp1mass.h

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// libCEED QFunctions for setting up BP1 mass operator
+
+// *****************************************************************************
+static int Fields(Ceed ceed, CeedQFunction *qf) {
+
+  // Create the Q-function that defines the action of the mass operator.
+  CeedQFunctionCreateInterior(ceed, 1, Mass, __FILE__ ":Mass", &qf);
+  CeedQFunctionAddInput(qf, "u", 1, CEED_EVAL_INTERP);
+  CeedQFunctionAddInput(qf, "rho", 1, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf, "v", 1, CEED_EVAL_INTERP);
+}
+// *****************************************************************************
+static int Mass(void *ctx, CeedInt Q,
```

Comment:
and this is the qfunction itself. Ideally I'd like to use this as a base and other backends (OCCA, CUDA, etc) can use this to build anything specific to that backend (like the OKL file).

Reply:
Ok this will work as long as the data accessed are of stride `Q`, if the layout is organized differently then something much more advanced has to be done on GPU.

Reply:
Everything should be loops in term of Q

Reply:
However, this prevents us from fusing operations on a GPU since this function is __global__, unfortunately making it a __device__ function doesn't resolve completely the problem, but might be closer to what we want. However, making the QFunction a __device__ function change the level at which the QFunction is described, indeed we cannot include the for loop on threads:
```
for (int i = blockIdx.x * blockDim.x + threadIdx.x;
    i < Q;
    i += blockDim.x * gridDim.x)
```
since the parallelization strategy is decided at a higher level (the level of the fused __global__ function), and cannot be at the __device__ function level.

Reply:
Ideally the user should have to only specify the names of the input/output variables (CeedScalar *rho = in[0], etc) and the action at the quadrature points (v[i] = rho[i] * u[i]). If a more complex loop control structure is required, we can work to adapt the code generated to best match the needs of OCCA or CUDA.